README.md
# Scinterm

Scinterm is a curses platform for [Scintilla][] that supports [ncurses][],
[PDCurses][], and X/Open Curses.

![Scinterm](images/scinterm.png)

It is highly recommended to run Scinterm in a UTF-8-aware terminal with a font
that supports many UTF-8 characters ("DejaVu Sans Mono" is one of them), since
Scinterm makes use of UTF-8 characters when drawing wrap symbols, some marker
symbols, and call tip arrows.

[Scintilla]: https://scintilla.org
[ncurses]: https://invisible-island.net/ncurses/
[PDCurses]: https://pdcurses.org

## Requirements

* Scinterm 3.1 requires Scintilla 4.2.0 - 4.x.
* Scinterm 3.0 requires Scintilla 3.20.0 - 3.21.0.
* Scinterm 2.0 requires Scintilla 3.20.0 - 3.21.0.
* Scinterm 1.12 requires Scintilla 3.11.0 - 3.11.2.
* Scinterm 1.11 requires Scintilla 3.10.0 - 3.10.6.
* Scinterm 1.10 requires Scintilla 3.8.0.
* Scinterm 1.9 requires Scintilla 3.7.5 - 3.7.6.
* Scinterm 1.8 requires Scintilla 3.6.3 - 3.7.4.
* Scinterm 1.7 requires Scintilla 3.6.3 - 3.7.4.
* Scinterm 1.6 requires Scintilla 3.5.5 - 3.6.2.
* Scinterm 1.5 requires Scintilla 3.5.2 - 3.5.4.
* Scinterm 1.4 requires Scintilla 3.5.0 - 3.5.1.
* Scinterm 1.3 requires Scintilla 3.4.2 - 3.4.4.
* Scinterm 1.2 requires Scintilla 3.3.7 - 3.4.1.
* Scinterm 1.1 requires Scintilla 3.2.2 - 3.3.6.
* Scinterm 1.0 requires Scintilla 3.2.2 - 3.3.6.

## Download

Scinterm releases can be found [here][].

[here]: https://github.com/orbitalquark/scinterm/releases

## Compile

After downloading Scinterm, it is recommended to unzip it into the top-level
directory of an instance of Scintilla, similar to other Scintilla platforms like
*gtk/* and *win32/*. After that, go into the Scinterm directory and run `make`
to build the usual *../bin/scintilla.a*.

You can optionally build the demo application, jinx, by going into *jinx/* and
running `make`. Pressing the `q` key quits the demo.

## Usage

Scinterm's Application Programming Interface [(API) documentation][] is located
in the project's *docs/* directory and covers how to create and interact with a
Scintilla widget in a terminal application.

[(API) documentation]: https://orbitalquark.github.io/scinterm/api.html

## Curses Compatibility

Scinterm lacks some Scintilla features due to the terminal's constraints:

* Any settings with alpha values are not supported.
* Autocompletion lists cannot show images (pixmap surfaces are not supported).
  Instead, they show the first character in the string passed to
  [`SCI_REGISTERIMAGE`][].
* Buffered and two-phase drawing is not supported.
* Caret settings like period, line style, and width are not supported
  (terminals use block carets with their own period definitions).
* Code pages other than UTF-8 have not been tested and it is possible some
  curses implementations do not support them.
* Drag and drop is not supported.
* Edge lines are not displayed properly (the line is drawn over by text lines).
* Extra ascent and descent for lines is not supported.
* Fold lines cannot be drawn above or below lines.
* Hotspot underlines are not drawn on mouse hover (`surface->FillRectangle()` is
  not supported).
* Indicators other than `INDIC_ROUNDBOX` and `INDIC_STRAIGHTBOX` are not drawn
  (`surface->LineTo()` and `surface->FillRectangle()` are not supported for
  drawing indicator shapes and pixmap surfaces are not supported). Translucent
  drawing and rounded corners are not supported either.
* Some complex marker types are not drawn properly or at all (pixmap surfaces
  are not supported and `surface->LineTo()` is not supported for drawing some
  marker shapes).
* Mouse cursor types are not supported.
* Up to 16 colors are supported, regardless of how many colors the terminal
  supports. They are (in "0xBBGGRR" format): black (`0x000000`), red
  (`0x000080`), green (`0x008000`), yellow (`0x008080`), blue (`0x800000`),
  magenta (`0x800080`), cyan (`0x808000`), white (`0xC0C0C0`), light black
  (`0x404040`), light red (`0x0000FF`), light green (`0x00FF00`), light yellow
  (`0x00FFFF`), light blue (`0xFF0000`), light magenta (`0xFF00FF`), light cyan
  (`0xFFFF00`), and light white (`0xFFFFFF`). Even if your terminal uses a
  different color map, you must use these color values with Scintilla;
  unrecognized colors are set to white by default. For some terminals, you may
  need to set a lexer style's `bold` attribute in order to use the light color
  variant.
* Some styles settings like font name, font size, and italic do not display
  properly (terminals use one only font, size and variant).
* X selections (primary and secondary) are not integrated into the clipboard.
* Zoom is not supported (terminal font size is fixed).
* When using the mouse in the Windows console, Shift+Double-click extends
  selections and quadruple-clicking inside a selection collapses it.

[`SCI_REGISTERIMAGE`]: https://scintilla.org/ScintillaDoc.html#SCI_REGISTERIMAGE

## Contribute

Scinterm is [open source][]. Feel free to report bugs and submit patches either
to the [mailing list][], or to me personally (mitchell.att.foicica.com). Thanks
to [everyone][] who has contributed.

[open source]: https://github.com/orbitalquark/scinterm
[mailing list]: https://foicica.com/lists
[everyone]: https://orbitalquark.github.io/scinterm/thanks.html


CHANGELOG.md
# Changelog

[Atom Feed][] | [PGP Public Key][]

[Atom Feed]: feed
[PGP Public Key]: https://foicica.com/foicica.pgp

## 2.0 (?)

Download:

* ?

Bugfixes:

* Fixed potential crashes with newer versions of ncurses.
* Fixed overlapping window drawing bug in Windows.
* Fixed drawing of block caret in overtype mode.

Changes:

* Update the cursor position when drawing focused views.
* Added userdata pointer for SCNotifications.
* Enable autoscrolling when clicking and dragging the mouse to make selections.
* [`scintilla_get_clipboard()`][] now returns a copy of clipboard text directly.

[`scintilla_get_clipboard()`]: api.html#scintilla_get_clipboard

## 1.9 (25 Jun 2017)

Download:

* [Scinterm 1.9][] | [PGP -- 1.9][]

Bugfixes:

* Fixed various compiler warnings.

Changes:

* Use a curses primitive instead of implementation-specific workaround when
  changing text attributes.
* Updated to [Scintilla][] 3.7.5.

[Scinterm 1.9]: download/scinterm_1.9.zip
[PGP -- 1.9]: download/scinterm_1.9.zip.asc
[Scintilla]: http://scintilla.org

## 1.8 (26 Apr 2016)

Download:

* [Scinterm 1.8][] | [PGP -- 1.8][]

Bugfixes:

* Many improvements to UTF-8 character handling.
* Fix building on some versions of Mac OSX.

Changes:

* Improved ncurses wide-character detection.
* Added [`scintilla_noutrefresh()`][] for more efficient refreshing.

[Scinterm 1.8]: download/scinterm_1.8.zip
[PGP -- 1.8]: download/scinterm_1.8.zip.asc
[Scintilla]: http://scintilla.org
[`scintilla_noutrefresh()`]: api.html#scintilla_noutrefresh

## 1.7 (03 Feb 2016)

Download:

* [Scinterm 1.7][] | [PGP -- 1.7][]

Bugfixes:

* None.

Changes:

* Delay creation of curses `WINDOW`s for as long as possible.
* Updated to [Scintilla][] 3.6.3.

[Scinterm 1.7]: download/scinterm_1.7.zip
[PGP -- 1.7]: download/scinterm_1.7.zip.asc
[Scintilla]: http://scintilla.org

## 1.6 (18 Apr 2015)

Download:

* [Scinterm 1.6][] | [PGP -- 1.6][]

Bugfixes:

* Fixed indicator drawing over multiple styles.

Changes:

* Updated to [Scintilla][] 3.5.5.

[Scinterm 1.6]: download/scinterm_1.6.zip
[PGP -- 1.6]: download/scinterm_1.6.zip.asc
[Scintilla]: http://scintilla.org

## 1.5 (10 Dec 2014)

Download:

* [Scinterm 1.5][] | [PGP -- 1.5][]

Bugfixes:

* Fixed positioning of autocomplete and call tips in some edge cases.
* Margins are no longer overwritten by long lines.
* Allow UTF-8 characters as list images.

Changes:

* Added complete mouse support.
* Added support for drawing call tip arrows.
* Implemented scrollbars.
* Added support for drawing tab arrows when visualizing whitespace.
* Added support for drawing wrap markers. (Requires a font with good
  glyph support, like DejaVu Sans Mono.)
* Added support for drawing nearly all line markers. (Requires a font with good
  glyph support, like DejaVu Sans Mono or Liberation Mono.)
* Added support for drawing indentation guides.
* Updated to [Scintilla][] 3.5.2.

[Scinterm 1.5]: download/scinterm_1.5.zip
[PGP -- 1.5]: download/scinterm_1.5.zip.asc
[Scintilla]: http://scintilla.org

## 1.4 (01 Sep 2014)

Download:

* [Scinterm 1.4][] | [PGP -- 1.4][]

Bugfixes:

* Fixed nested folding marker drawing.

Changes:

* Updated to [Scintilla][] 3.5.0.

[Scinterm 1.4]: download/scinterm_1.4.zip
[PGP -- 1.4]: download/scinterm_1.4.zip.asc
[Scintilla]: http://scintilla.org

## 1.3 (09 Jul 2014)

Download:

* [Scinterm 1.3][] | [PGP -- 1.3][]

Bugfixes:

* Clip long lines properly when view width is less than `COLS`.
* Fixed painting bug in resized views.

Changes:

* Updated to [Scintilla][] 3.4.2.

[Scinterm 1.3]: download/scinterm_1.3.zip
[PGP -- 1.3]: download/scinterm_1.3.zip.asc
[Scintilla]: http://scintilla.org

## 1.2 (05 Feb 2014)

Download:

* [Scinterm 1.2][] | [PGP -- 1.2][]

Bugfixes:

* Fixed bug with calltip styles.
* Fixed various compile issues and warnings.
* Interpret UTF-8 input correctly.
* Fix crash when trying to draw a call tip larger than the window.
* Fixed bug in nested fold markers not showing up.

Changes:

* Added support for `INDIC_ROUNDBOX` and `INDIC_STRAIGHTBOX`, but with no
  translucency.
* Support curses, PDCurses, and X/Open Curses.
* Recognize up to 16 separate colors for terminals that support it.
* Updated to [Scintilla][] 3.3.7.

[Scinterm 1.2]: download/scinterm_1.2.zip
[PGP -- 1.2]: download/scinterm_1.2.zip.asc
[Scintilla]: http://scintilla.org

## 1.1 (11 Dec 2012)

Download:

* [Scinterm 1.1][] | [PGP -- 1.1][]

Bugfixes:

* Fixed crash when pasting uninitialized clipboard text.
* Fixed bug with drawing around italic characters at EOLs.
* Display control characters properly.
* Fixed bug in searching autocompletion lists.

Changes:

* Expose [`SCI_COLOR_PAIR()`][] macro.

[Scinterm 1.1]: download/scinterm_1.1.zip
[PGP -- 1.1]: download/scinterm_1.1.zip.asc
[`SCI_COLOR_PAIR()`]: api.html#SCI_COLOR_PAIR

## 1.0 (31 Aug 2012)

Download:

* [Scinterm 1.0][] | [PGP -- 1.0][]

Initial release.

[Scinterm 1.0]: download/scinterm_1.0.zip
[PGP -- 1.0]: download/scinterm_1.0.zip.asc


jinx/jinx.c
// Copyright 2012-2020 Mitchell mitchell.att.foicica.com. See LICENSE.

#include <locale.h>
#include <sys/time.h>
#include <curses.h>

#include "Scintilla.h"
#include "SciLexer.h"
#include "ScintillaCurses.h"

#define SSM(m, w, l) scintilla_send_message(sci, m, w, l)

typedef void Scintilla;

void scnotification(Scintilla *view, int msg, SCNotification *n, void *userdata)
{
  //printw("SCNotification received: %i", n->nmhdr.code);
}

int main(int argc, char **argv) {
  setlocale(LC_CTYPE, ""); // for displaying UTF-8 characters properly
  initscr(), raw(), cbreak(), noecho(), start_color();
  Scintilla *sci = scintilla_new(scnotification, NULL);
  curs_set(0); // Scintilla draws its own cursor

  SSM(SCI_STYLESETFORE, STYLE_DEFAULT, 0xFFFFFF);
  SSM(SCI_STYLESETBACK, STYLE_DEFAULT, 0);
  SSM(SCI_STYLECLEARALL, 0, 0);
  SSM(SCI_SETLEXER, SCLEX_CPP, 0);
  SSM(SCI_SETKEYWORDS, 0, (sptr_t)"int char");
  SSM(SCI_STYLESETFORE, SCE_C_COMMENT, 0x00FF00);
  SSM(SCI_STYLESETFORE, SCE_C_COMMENTLINE, 0x00FF00);
  SSM(SCI_STYLESETFORE, SCE_C_NUMBER, 0xFFFF00);
  SSM(SCI_STYLESETFORE, SCE_C_WORD, 0xFF0000);
  SSM(SCI_STYLESETFORE, SCE_C_STRING, 0xFF00FF);
  SSM(SCI_STYLESETBOLD, SCE_C_OPERATOR, 1);
  SSM(SCI_INSERTTEXT, 0, (sptr_t)
      "int main(int argc, char **argv) {\n"
      "    // Start up the gnome\n"
      "    gnome_init(\"stest\", \"1.0\", argc, argv);\n}");
  SSM(SCI_SETPROPERTY, (uptr_t)"fold", (sptr_t)"1");
  SSM(SCI_SETMARGINWIDTHN, 0, 2);
  SSM(SCI_SETMARGINWIDTHN, 2, 1);
  SSM(SCI_SETMARGINMASKN, 2, SC_MASK_FOLDERS);
  SSM(SCI_SETMARGINSENSITIVEN, 2, 1);
  SSM(SCI_SETAUTOMATICFOLD, SC_AUTOMATICFOLD_CLICK, 0);
  SSM(SCI_SETFOCUS, 1, 0);
  scintilla_refresh(sci);

  printf("\033[?1000h"); // enable mouse press and release events
  //printf("\033[?1002h"); // enable mouse press, drag, and release events
  //printf("\033[?1003h"); // enable mouse move, press, drag, and release events
  mousemask(ALL_MOUSE_EVENTS, NULL);
  mouseinterval(0);

  // Non-UTF8 input.
  int c = 0;
  MEVENT mouse;
  WINDOW *win = scintilla_get_window(sci);
  while ((c = wgetch(win)) != 'q') {
    if (c != KEY_MOUSE) {
      if (c == KEY_UP) c = SCK_UP;
      else if (c == KEY_DOWN) c = SCK_DOWN;
      else if (c == KEY_LEFT) c = SCK_LEFT;
      else if (c == KEY_RIGHT) c = SCK_RIGHT;
      scintilla_send_key(sci, c, FALSE, FALSE, FALSE);
    } else if (getmouse(&mouse) == OK) {
      int event = SCM_DRAG, button = 0;
      if (mouse.bstate & BUTTON1_PRESSED)
        event = SCM_PRESS, button = 1;
      else if (mouse.bstate & BUTTON1_RELEASED)
        event = SCM_RELEASE, button = 1;
      struct timeval time = {0, 0};
      gettimeofday(&time, NULL);
      int millis = time.tv_sec * 1000 + time.tv_usec / 1000;
      scintilla_send_mouse(sci, event, millis, button, mouse.y, mouse.x,
                           mouse.bstate & BUTTON_SHIFT,
                           mouse.bstate & BUTTON_CTRL,
                           mouse.bstate & BUTTON_ALT);
    }
    scintilla_refresh(sci);
  }
  // UTF-8 input.
  //SSM(SCI_SETCODEPAGE, SC_CP_UTF8, 0);
  //wint_t c = {0};
  //WINDOW *win = scintilla_get_window(sci);
  //while (c != 'q') {
  //  int status = wget_wch(win, &c);
  //  if (status == ERR)
  //    continue;
  //  else if (status == KEY_CODE_YES) {
  //    if (c == KEY_UP) c = SCK_UP;
  //    else if (c == KEY_DOWN) c = SCK_DOWN;
  //    else if (c == KEY_LEFT) c = SCK_LEFT;
  //    else if (c == KEY_RIGHT) c = SCK_RIGHT;
  //  }
  //  scintilla_send_key(sci, c, FALSE, FALSE, FALSE);
  //  scintilla_refresh(sci);
  //}

  printf("\033[?1000l"); // disable mouse press and release events
  //printf("\033[?1002l"); // disable mouse press, drag, and release events
  //printf("\033[?1003l"); // disable mouse move, press, drag, and release

  scintilla_delete(sci);
  endwin();

  return 0;
}
